src/datomish/db.cljc

```diff
@@ -12,31 +12,599 @@
+   })
+
+(def ^{:private true} bootstrap-idents
+  {:db/ident             1
+   :db.part/db           2
+   :db/txInstant         3
+   :db.install/partition 4
+   :db.install/valueType 5
+   :db.install/attribute 6
+   :db/valueType         7
+   :db/cardinality       8
+   :db/unique            9
+   :db/isComponent       10
+   :db/index             11
+   :db/fulltext          12
+   :db/noHistory         13
+   :db/add               14
+   :db/retract           15
+   :db.part/tx           16
+   :db.part/user         17
+   :db/excise            18
+   :db.excise/attrs      19
+   :db.excise/beforeT    20
+   :db.excise/before     21
+   :db.alter/attribute   22
+   :db.type/ref          23
+   :db.type/keyword      24
+   :db.type/integer      25 ;; TODO: :db.type/long, to match Datomic?
+   :db.type/string       26
+   :db.type/boolean      27
+   :db.type/instant      28
+   :db.type/bytes        29
+   :db.cardinality/one   30
+   :db.cardinality/many  31
+   :db.unique/value      32
+   :db.unique/identity   33})
+
+(defn- bootstrap-tx-data []
+  (concat
+    (map (fn [[ident entid]] [:db/add entid :db/ident ident]) bootstrap-idents)
+    (map (fn [[ident attrs]] (assoc attrs :db/id ident)) bootstrap-symbolic-schema)
+    (map (fn [[ident attrs]] [:db/add :db.part/db :db.install/attribute (get bootstrap-idents ident)]) bootstrap-symbolic-schema) ;; TODO: fail if nil.
+    ))
+
+(defn <idents [sqlite-connection]
+  "Read the ident map materialized view from the given SQLite store.
+  Returns a map (keyword ident) -> (integer entid), like {:db/ident 0}."
+
+  (let [<-SQLite (get-in ds/value-type-map [:db.type/keyword :<-SQLite])] ;; TODO: make this a protocol.
+    (go-pair
+      (let [rows (<? (->>
+                       {:select [:ident :entid] :from [:idents]}
+                       (sql/format)
+                       (s/all-rows sqlite-connection)))]
+        (into {} (map (fn [row] [(<-SQLite (:ident row)) (:entid row)])) rows)))))
+
+(defn <current-tx [sqlite-connection]
+  "Find the largest tx written to the SQLite store.
+  Returns an integer, -1 if no transactions have been written yet."
+
   (go-pair
-    (when-not (= sqlite-schema/current-version (<? (sqlite-schema/<ensure-current-version sqlite-connection)))
-      (raise-str "Could not ensure current SQLite schema version."))
-    (->DB sqlite-connection)))
+    (let [rows (<? (s/all-rows sqlite-connection ["SELECT COALESCE(MAX(tx), -1) AS current_tx FROM transactions"]))]
+      (:current_tx (first rows)))))
+
+(defn <symbolic-schema [sqlite-connection]
+  "Read the schema map materialized view from the given SQLite store.
+  Returns a map (keyword ident) -> (map (keyword attribute -> keyword value)), like
+  {:db/ident {:db/cardinality :db.cardinality/one}}."
+
+  (let [<-SQLite (get-in ds/value-type-map [:db.type/keyword :<-SQLite])] ;; TODO: make this a protocol.
+    (go-pair
+      (->>
+        (->>
+          {:select [:ident :attr :value] :from [:schema]}
```

Comment:
Yeah, the idea is that it might make the `group-by` more efficient, because you get an ordered walk. Don't know if that'll happen in practice.